Improve Territory Name Placement

ron-murhammer opened the pull request:
Instead of just finding the bounding rectangle and placing the name in
the center, actually try to find the best location within the territory.
This now finds the rectangle that can fit the name, that is the closest
to the vertical center, and has a large width at that location. If one
isn't found then default back to the bounding rectangle center.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/triplea-game/triplea/622)
<!-- Reviewable:end -->

DanVanAtta commented on src/games/strategy/triplea/ui/MapData.java:
```diff
@@ -700,6 +701,50 @@ private Rectangle getBoundingRectWithTranslate(final List<Polygon> polys, final
+   * that location. If there isn't any rectangles that can fit the name then default back to the
+   * bounding rectangle.
+   */
+  public Rectangle getBestTerritoryNameRect(final Territory t, final FontMetrics fm) {
```
pls spell out `t` and `fm`

DanVanAtta replied:
Mehod is in the wrong location. MapData should not know anything about placing names in rectangles and or polygons.

DanVanAtta replied:
Missing test for this function.

ron-murhammer replied:
Updated variable names and moved method. This is UI code which really can't be unit tested well.

DanVanAtta replied:
Why not easily tested?

ron-murhammer replied:
Because the purpose of the method is to try to better visually place the territory names. Hard to write a test to check how well a name is visually placed. Imagine even trying to do TDD for this. What tests would you write to prove the method is working as intended? Besides a basic test that shows you actually return a rectangle there isn't much.

DanVanAtta commented on src/games/strategy/triplea/ui/MapData.java:
```diff
@@ -700,6 +701,50 @@ private Rectangle getBoundingRectWithTranslate(final List<Polygon> polys, final
+   */
+  public Rectangle getBestTerritoryNameRect(final Territory t, final FontMetrics fm) {
+    final Rectangle territoryBounds = getBoundingRect(t);
+    Rectangle maxRectangle = territoryBounds;
```
Pls place `maxRectangle`, the object you are returning, first. You have two Rectangle objects on the first line, it was a bit of a guess which one was the return object.

ron-murhammer replied:
I'd like to leave territoryBounds first since that is what the function is returning. I've renamed the return variable to 'result'.

DanVanAtta replied:
` return maxRectangle;`

Just going by what the code says.

ron-murhammer replied:
This is already updated. You must be looking at an outdated diff...

DanVanAtta commented on src/games/strategy/triplea/ui/MapData.java:
```diff
@@ -700,6 +701,50 @@ private Rectangle getBoundingRectWithTranslate(final List<Polygon> polys, final
+  public Rectangle getBestTerritoryNameRect(final Territory t, final FontMetrics fm) {
+    final Rectangle territoryBounds = getBoundingRect(t);
+    Rectangle maxRectangle = territoryBounds;
+    final int minX = territoryBounds.x;
```
Please inline `minX` and `minY`.
There are too many variables declared to be remembered at once. This is bad for being able to understand things easily when you have to look up everything because your short term memory is out of space. There are 11 variables declared.. The only way to remember them is to figure out which ones can be ignored. This is work that should be done for the reader.

ron-murhammer replied:
Inlined. Its about a wash I think but I'm fine either way.

DanVanAtta replied:
I caution that you may have too much context, you know that the variables really don't matter. When you approach this six months from now, and try to keep 11 different variables in your head before you hit the triple for loop...

DanVanAtta commented on src/games/strategy/triplea/ui/MapData.java:
```diff
@@ -700,6 +701,50 @@ private Rectangle getBoundingRectWithTranslate(final List<Polygon> polys, final
+    final int maxX = minX + territoryBounds.width;
+    final int maxY = minY + territoryBounds.height;
+    final int centerY = minY + territoryBounds.height / 2;
+    final List<Polygon> polygons = m_polys.get(t.getName());
```
Inline this one as well. It's inner loop code, but savings is  on the order of nanoseconds.

DanVanAtta replied:
Please convert to m_polys to a `MultiMap<String,Polygon>`. It will fix the NPE potential below and simplify things.

ron-murhammer replied:
Inlined. Not looking to change variable types outside this method especially since its not moved out of this class.

DanVanAtta replied:
Please push the update. Why not change to a multimap? The conversion is 1:1, and will avoids a lot of possible NPEs and NPE checking.

ron-murhammer replied:
Update pushed. I don't want to start changing variables outside of the core changes I'm making. So out of scope.

DanVanAtta commented on src/games/strategy/triplea/ui/MapData.java:
```diff
@@ -700,6 +701,50 @@ private Rectangle getBoundingRectWithTranslate(final List<Polygon> polys, final
+    final int centerY = minY + territoryBounds.height / 2;
+    final List<Polygon> polygons = m_polys.get(t.getName());
+    final int minWidth = fm.stringWidth(t.getName());
+    final int increment = fm.getAscent();
```
`fm.getAscent()` seems like vertical height, yet we increment both startX and endX by the increment. Can you help clear that up?

DanVanAtta replied:
It seems like `fm.getAscent()` would be jumping up by the height of the font. Why do we increment in by that amount in the horizontal direction? In fact, why by that amount in the vertical? If we are searching for the best location, then incrementing by one would guarantee we don't jump past a good spot.

ron-murhammer replied:
Correct. Its used both as the vertical and horizontal increment.

Performance. I invite you to try changing 'increment' to '1' for all the loops and launch the game. Essentially this is an approximation of the best location using the font height as the increment to create a grid to try placing the name. It probably works about 98% as effective as if you used '1' as the increment. If you want to use an increment of '1' you would need to have a much more complex algorithm so performance would be reasonable.

DanVanAtta replied:
What is the runtime performance of this when the increment is equal `1`? VS when increment is used? All in-memory operations, should be pretty fast in both cases.

Is `fm.getAscent` impacted at all by system font settings? For example, if someone has font set to 36 point, will that make the increment become very large? Seems like the incremetn should be a fraction of the total height, IE: Math.cieling( height / incrementSlices), where incrementSlices is equal to something like 10, 20 or 50.

DanVanAtta commented on src/games/strategy/triplea/ui/MapData.java:
```diff
@@ -700,6 +701,50 @@ private Rectangle getBoundingRectWithTranslate(final List<Polygon> polys, final
+    final int minWidth = fm.stringWidth(t.getName());
+    final int increment = fm.getAscent();
+    int maxScore = 0;
+    for (int startX = minX; startX < maxX - increment; startX += increment) {
```
Some commentary on the for loops would go a long way. The triple for loop is a code smell, deep nesting, and has high cyclomatic complexity.

ron-murhammer replied:
Added some comments. Triple loop is necessary to test all the possibilities so not a code smell here. Unless you move to a different algorithm and potentially cache some data, its necessary.

DanVanAtta commented on src/games/strategy/triplea/ui/MapData.java:
```diff
@@ -700,6 +701,50 @@ private Rectangle getBoundingRectWithTranslate(final List<Polygon> polys, final
+      for (int startY = minY; startY < maxY - increment; startY += increment) {
+        for (int endX = maxX; endX > startX; endX -= increment) {
+          final Rectangle current = new Rectangle(startX, startY, endX - startX, increment);
+          final int verticalFromEdge = territoryBounds.height / 2 - Math.abs(centerY - (startY + increment));
```
`verticalFromEdge` - which edge?

ron-murhammer replied:
Either edge (top or bottom). This gives a higher value when near the vertical center which is then multiplied by the max width at that vertical to give 'score'.

DanVanAtta replied:
Pls rename variable so it is clear.

ron-murhammer replied:
Suggestions? verticalFromEdge makes sense to me but I'm open to better names.

DanVanAtta replied:
Perhaps: `distanceFromVerticalCenter`

DanVanAtta commented on src/games/strategy/triplea/ui/screen/IDrawable.java:
```diff
@@ -128,26 +128,29 @@ public void draw(final Rectangle bounds, final GameData data, final Graphics2D g
-    // otherwise, put it in the center
     final Point namePlace = mapData.getNamePlacementPoint(territory);
     if (namePlace == null) {
+      final Rectangle territoryBounds = mapData.getBestTerritoryNameRect(territory, fm);
```
`mapData.getBoundingRect(territory);` is called in other locations, it's a bit odd this is the only spot that benefits. If that is really the case, why not make this a private method attached to this class that uses it, and have it accept a mapData object?

ron-murhammer replied:
Done.

DanVanAtta commented on src/games/strategy/triplea/ui/MapData.java:
```diff
@@ -700,6 +701,50 @@ private Rectangle getBoundingRectWithTranslate(final List<Polygon> polys, final
+    final int increment = fm.getAscent();
+    int maxScore = 0;
+    for (int startX = minX; startX < maxX - increment; startX += increment) {
+      for (int startY = minY; startY < maxY - increment; startY += increment) {
```
'startY' as a loop variable does not make much sense. Sounds like the position at the start of the loop, but it's instead the current position. Same for `endX` + `startX`

ron-murhammer replied:
Renamed to left, right, top instead of start and end. Was referring to start of rectangle not start of loop.

DanVanAtta commented on src/games/strategy/triplea/ui/MapData.java:
```diff
@@ -700,6 +701,50 @@ private Rectangle getBoundingRectWithTranslate(final List<Polygon> polys, final
+          if (current.width > minWidth && score > maxScore) {
+            boolean isContained = false;
+            for (final Polygon polygon : polygons) {
+              if (polygon.contains(current)) {
```
Looks like you can exit the first time you find a fit.

The Math.abs function is effectivey:

`Math.abs(minY + territoryBounds.height / 2 - startY - increment))`
The only non-constant is the Y value, which will only grow. Which will only make this value shrink with each successive iteration: `final int verticalFromEdge = territoryBounds.height / 2 - Math.abs(centerY - (startY + increment));`

Same thing for the X direction, the function that adjusts your score is linear, it will only shrink with successive loop iterations.

ron-murhammer replied:
Not sure exactly what your point or question is here.

verticalFromEdge will start at its min value then moves to max at vertical center then move back to min value (ex. 1 2 3 2 1) and yes is only influenced by Y.

width is influenced by both left X and right X (1st and 3rd loops).

The idea is to find the rectangle that is the widest and closest to vertical center.

A 'break' is used in the inner most loop since it loops backwards starting with the widest rectangle at the given point (X, Y) so as soon as you get a success then you don't need to check any other rectangles at that point since they will all have a lower score.

DanVanAtta replied:
The inner most loop is only dependent on Y (`startY`), in:
` final int verticalFromEdge = territoryBounds.height / 2 - Math.abs(centerY - (startY + increment));`

As Y grows, the value of `veritcalFromEdge` only shrinks  until  this is true: `startY+increment > 2 * centerY`
It does appear we would be out of the rectangle if that were ever true.
This means that the first iteration of the Y loop will find a max value, and same for the X loops.

ron-murhammer replied:
No, please re-read the above comment. verticalFromEdge is at a max when Y is halfway through the loop. X always starts at the max but that doesn't mean the rectangle will be contained in the territory and is why once its found a break is used.

DanVanAtta replied:
What is that max value. Please indicate it in the code so it's clear which ranges of values we will have in the Math.abs function.

If verticalFromEdge is at the max when Y is half way, then start your loop there. There are two linear functions, and an absolute function, which you should be able to treat as linear. Given it's all linear, you should be able to start with the max value and get this down to `O(n)`

ron-murhammer commented:
@DanVanAtta I think you were still looking at an outdated diff. Please check again.

DanVanAtta commented:
The new function still needs a test.

DanVanAtta commented on src/games/strategy/triplea/ui/screen/IDrawable.java:
```diff
@@ -1,5 +1,19 @@
+import games.strategy.triplea.image.TileImageFactory;
+import games.strategy.triplea.ui.IUIContext;
+import games.strategy.triplea.ui.MapData;
+import games.strategy.triplea.util.Stopwatch;
```
unused import

ron-murhammer replied:
Not sure what you are referring to as StopWatch is used.

DanVanAtta replied:
Thought we had killed off the StopWatch usages. You're right there is something that is being timed with it.

DanVanAtta commented on src/games/strategy/triplea/ui/screen/IDrawable.java:
```diff
@@ -128,26 +128,29 @@ public void draw(final Rectangle bounds, final GameData data, final Graphics2D g
     }
-    final Rectangle territoryBounds = mapData.getBoundingRect(territory);
+
+    // get font metrics
```
unnecessary comment. We have the word 'getFontMetrics()' three lines later, comment is repeating the code.

ron-murhammer replied:
Comment removed.

DanVanAtta commented:
What is the timing difference between increment vs using 1? How large can a polygon be before this algorithm starts to slow down? `O(n^3)` seems far too slow, given linear functions, it does not need to be that slow.

DanVanAtta commented on src/games/strategy/triplea/ui/screen/IDrawable.java:
```diff
@@ -128,26 +128,29 @@ public void draw(final Rectangle bounds, final GameData data, final Graphics2D g
-      y += (int) territoryBounds.getHeight() >> 1;
-      x -= fm.stringWidth(territory.getName()) >> 1;
-      y += fm.getAscent() >> 1;
+      x += (int) territoryBounds.getWidth() / 2;
```
Why do this increment of `x` on abrand new line? Pls just inline it with line 143.

ron-murhammer replied:
No idea as I didn't write. Merged into one line for x and y.

DanVanAtta commented on src/games/strategy/triplea/ui/screen/IDrawable.java:
```diff
@@ -180,6 +183,57 @@ public void draw(final Rectangle bounds, final GameData data, final Graphics2D g
+
+    // Loop through the grid moving the starting point and determining max width at that point
+    for (int leftX = territoryBounds.x; leftX < maxX - increment; leftX += increment) {
+      for (int topY = territoryBounds.y; topY < maxY - increment; topY += increment) {
```
You mentioned the score is maximized when Y is half way through the loop. That should be made clear in the code. Right now a person has to unwind all variables back to their `territoryBounds` definitions, which gets pretty gnarly very readily.

DanVanAtta replied:
If we find a maximal value half way through the Y loop, we can break from that loop, and we can break from the outer most loop as well.

ron-murhammer replied:
How do you know its a maximal value?

DanVanAtta commented on src/games/strategy/triplea/ui/screen/IDrawable.java:
```diff
@@ -180,6 +183,57 @@ public void draw(final Rectangle bounds, final GameData data, final Graphics2D g
+
+          // Find current rectangle and the 'score' based on how close to vertical center and width
+          final Rectangle current = new Rectangle(leftX, topY, rightX - leftX, increment);
+          final int verticalFromEdge = territoryBounds.height / 2 - Math.abs(centerY - (topY + increment));
```
The `abs()` condition should just use different variables, or be written back in terms of territoryBounds so it can be understood more simply.

ron-murhammer replied:
Pretty sure that would make it even harder to understand.

ron-murhammer commented:
@DanVanAtta Added a few clarifications.

Feel free to optimize performance if you'd like but i tested with WaW (one of the largest maps with territory name text) and performance is right at the point of starting to make things slower. I can tell you with the current algorithm you definitely don't want to use 1 and also get very little benefit from doing so over font size. This is run for every territory so its not just size of a single territory that matters, its the size of the map. It isn't O(n^3) and not even sure what your 'n' stands for. It is O(n^2*m) where n=(width/font size) and m=(height/font size). The slow portion is also `polygon.contains(current)` which is often avoided by checking width and score first which are quick calculations.

Honestly, I would recommend taking a step back and better understanding the problem that is being addressed here and then read through the code. You still don't seem to fully understand the change or how it works. Also I recommend you do some functional testing since this is a UI change so we can refactor the code as much as possible but it is doesn't actually improve functionality then there is no point. You can also quickly adjust the increment to see the effect on performance. I don't think the code is going to get much more readable then it is now and I believe the change improves or is equal to existing functionality in all cases. So would like to merge this and we can always look to optimize it later on.

DanVanAtta commented:
" I would recommend taking a step back and better understanding the problem that is being addressed here and then read through the code"

I think I do. The problem is that the code requires a TON of work to understand.This is adding to maintenance burden, this is complicated enough nobody will want to ever touch it in fear it breaks something.

Pls Add some comments to document the logic. Specifically the min/max range of the Math.abs function. I recommend extracting all of the interior loop functionality to helper functions so the logic can stay at a very high level.

Add a test for this code pls.

DanVanAtta commented:
" performance is right at the point of starting to make things slower."
This is *very* concerning. We can have ever larger maps, and some folks have weaker hardware. You mentioned caching, it looks like it could be applied cleanly here to avoid re-calcing the values.

DanVanAtta commented on src/games/strategy/triplea/ui/screen/IDrawable.java:
```diff
@@ -128,26 +128,24 @@ public void draw(final Rectangle bounds, final GameData data, final Graphics2D g
-      x -= fm.stringWidth(territory.getName()) >> 1;
-      y += fm.getAscent() >> 1;
+      final Rectangle territoryBounds = getBestTerritoryNameRect(mapData, territory, fm);
+      x = territoryBounds.x + (int) territoryBounds.getWidth() / 2 - fm.stringWidth(territory.getName()) / 2;
```
Is this any better with parens to explicitly call out the order of operations?

ron-murhammer replied:
Could be argued either way. I think its probably better without but don't really feel strongly either way.

ron-murhammer commented:
@DanVanAtta Maybe that wasn't worded well but essentially I tried some smaller increments and got noticeably worse performance when starting a game. With this increment there is essentially no difference. It isn't multi-threaded and doesn't rely on high memory usage so should be relatively similar performance across machines (I'm testing on a decent laptop but nothing crazy high end). Obviously without releasing this to a larger test group its hard to say. I also doubt many maps will be much larger than WaW in terms of size and number of territories. 

Adding any caching then brings in memory concerns and makes the algorithm you are already struggling to understand much more complex. I'd rather keep it as simple as possible it is works well already. We also have plenty of caching and memory problems that exist today (map caching, etc). The saying usually goes 'don't optimize til you have to'.

I looked at extracting into some helper methods but it really doesn't help much do to the sheer number of variables. You end up with helper functions with 10 parameters.

I added a few more comments but I really don't think many more are going to help much.

DanVanAtta commented:
- memory concerns are pretty negligible for caching. We are already storing data about every polygon, and we have images for each one. Compare that to storing just a single number per polygon. A cache would consume less than 1k easily. The complexity of a cache here could be low, particularly if it is done right. At the end of the day we do have a performance requirement which overrules code complexity concerns (I present the NewGameChooserModel cache as an example)

- What is the timing of the method for you when running WaW?

- The method is pure logic - should have a test. I'd rather review a test that will deliver value over time rather than test it the once myself. 

- helper functions don't have to take every parameter. Perhaps instead of extracting labels so early, pass the high level objects around, something like this: 
```
for (int x = start; x < end; x += increment) {
   if( fits( x, rectangle, polygon )) {
       return getBoundingRectangle( x, rectangle, polygon)
    }
}

fits( int xPosition, rectangle, polygon ) {
   for( int y = 0; ...)
}
```

ron-murhammer commented:
@DanVanAtta 
- What are you suggesting we cache? Why would we cache 1 number per polygon?
- Didn't time it. Feel free to. 
- Feel free to add a test or explain what you would make input/output for such a test
- I really don't think helper functions are gonna help in this case especially at that level. Due to all the variables its easier to keep it in one place.
- Your sudo code doesn't really make sense. It is doing way more than checking if it fits. And why would you return bounding rectangle if it fits? Honestly not sure where you are even going with it. Still not sure you really understand it.

DanVanAtta commented on src/games/strategy/triplea/ui/screen/IDrawable.java:
```diff
@@ -180,6 +178,59 @@ public void draw(final Rectangle bounds, final GameData data, final Graphics2D g
+          if (current.width > minWidth && score > maxScore) {
+
+            // Check to make sure rectangle is contained in the territory
+            boolean isContained = false;
```
This is a good place to insert a helper function.
`boolean isContained = isRectangleContainedInTerritory(current, polygons);`

ron-murhammer replied:
Yep this is about the only place you can add a reasonable helper function. I'm fine doing that but it really doesn't make the method any easier to understand as the difficult part is the loops and scoring.

DanVanAtta commented:
Obviously the return value of the function is the thing to cache. Are we splitting hairs? 1 number vs 4. This is your PR, it's up to you to show that the performance of the game will not be negatively impacted. 

For test - likely it's easiest to use strategy pattern to test the logic piece independent of the loops. As is, give mapData one polygon and control when it returns true for: `(polygon.contains(current)) {`. Then from there you only need to control the values of the territory bounds. The font should be easy to mock out as well which would let you control the increment size. 

You never answered the question about system fonts, if someone has accessibility sized fonts on, will that cause the increment size to be too large? Why not instead calculate a fraction of the total height and always do a known number of iterations rather than relying on font size for the increment? Why does using vertical font size make sense for doing the horizontal increment? Shouldn't the two values, horizontal and vertical increments likely be different?

DanVanAtta commented:
Also - still not addressed, why not break out of the outermost X loop when you have found a max score? 

From these three lines:
```
201. for (int x = territoryBounds.x; x < maxX - increment; x += increment) {
204. final Rectangle current = new Rectangle(x, y, endX - x, increment);
210. final int score = verticalDistanceFromEdge * current.width;
```

`Width = endX - x`; As X increases, then score will decrease. It does seem like there can also be an early termination on the outer most loop.

ron-murhammer commented:
@DanVanAtta Based on your comments you still don't seem to understand the functionality or code.

- Why would you cache the return value? This method is only called once per territory so caching the return value would be absolutely useless.
- So for every PR we have to somehow show performance isn't negatively impacted? What does 'negatively impacted' exactly mean? You seem to be coming up with new PR requirements that have no real definition. No matter what you do this process is going to take some real time to calculate all the territory name placement for a larger map. There isn't a simple solution unless you go back to what we already had that just put the name in the center of the bounding rectangle...
- If you increase the font then the increment will increase but it should still do a decent job of placing the name. Again at some point the small territories just can't fit the name especially if you increase the font size. You could use a fraction of the height but then you are potentially checking the same number of points for every territory when some territories can be 100x larger than others so probably should test more points. Best way to check this would be actually try increasing the font to a larger size say 24 and see if that makes the territory name placements considerably worse on any maps. There is also a performance consideration where using the font size makes the overall performance based mostly on size of the map where basing the increment on height/width would have the performance be more impacted by number of territories.
- Font size is just a decent number to start from and most likely whatever your vertical increment is, your horizontal increment should be similar. Its essentially about creating some sort of grid to test starting points on and the boxes on the grid should be close to squares. so either the same value or probably based on the width.
- How would you know when you found the max score since its also based on 'verticalDistanceFromEdge'?

Maybe you should read some articles on finding max rectangles inside polygons as essentially this algorithm is closely related to that:
http://www.drdobbs.com/database/the-maximal-rectangle-problem/184410529
http://d3plus.org/blog/behind-the-scenes/2014/07/08/largest-rect/

ron-murhammer commented:
@DanVanAtta Also now that we are at around 65 comments for what boils down to about 50 lines of code, it probably would be a good idea for us both to look at code review best practices such as: http://www.developer.com/tech/article.php/3579756/Effective-Code-Reviews-Without-the-Pain.htm

I think we both have deviated some from those. Might be worth adding a few good links to our coding standard wiki at one of these pages:
https://github.com/triplea-game/triplea/wiki/Code-Guidelines
https://github.com/triplea-game/triplea/wiki/Code-Reviewers-and-Pull-Request-Guidelines

DanVanAtta commented:
> "What would caching the return value gain us? "

You had mentioned it early on??


> "This method is only called once per territory so caching the return value wouldn't do anything unless you want to cache it across games."

There is code to unload map tiles to conserve memory. If someone were to scroll a lot, would they potentially hit a lot of calculation?


> So for every PR we have to somehow show performance isn't negatively impacted? What does 'negatively impacted' exactly mean? 

If there is concern that performance could be a problem, then yes, you need to show it's not. I'll give the branch a thorough test. 'negatively impacted' in our context means slow enough that you notice when on weak hardware. It's subjective, but effectively 'noticeable' is the bar. 


> You seem to be coming up with new PR requirements that have no concrete definition. 

I don't think I'm making up the requirement of keeping the game playable. I'll mention this once, the links you said to keep the comments about the code. My comments have been that:
A) there is a requirement of the code to run relatively fast so the game UX stays good
B) This code has a triple for loop with calculations inside, and inside of UI drawing code that could maybe get called a lot. 

Meanwhile you've been disparaging my understanding, and accusing me of having a shifting bar.

> I did some basic performance testing on WaW as its one of the largest maps and everything works well for me. No matter what you do this process is going to take some real time to calculate all the territory name placement for a larger map. There isn't a simple solution unless you go back to what we already had that just put the name in the center of the bounding rectangle.

Pls double check the block is executed as many times as you think it is. If the data is calc'ed only once, and takes far less than a second (noticeable pause), then it's fine.

> If you increase the font then the increment will increase but it should still do a decent job of placing the name. Again at some point the small territories just can't fit the name especially if you increase the font size. You could use a fraction of the height but then you are potentially checking the same number of points for every territory when some territories can be 100x larger than others so probably should test more points. 

If someone has their font small, could we evaluate an excessive amount of data points? Are there any good test case territories on WaW to look at where increment has an effect? Have you experimented with half font size?


> How would you know when you found the max score since its also based on 'verticalDistanceFromEdge'?


Three loops:
- x left hand edge that moves to the right
- y that moves from bottom to top
- x right hand edge that moves right to left

`verticalDistanceFromEdge` only depends on the second loop. If you find a 'y' value that fits, the next 'x left hand edge' value will only be for a smaller width. 

Either that, or there is some dependency on the 'x left hand' that I have missed. If so, pls update the code to make that dependency to be obvious.


Bottom line:
The article you linked mentioned that code quality and maintainability were a goal. The initial code submitted required too much effort. A person shouldn't have to work like hell to understand small sections of code, there is *way* too much code to go through to get bogged down so much. 

At this point though with all of the changes things are getting to be much more readily comprehensible. The one place you can add a helper function, pls do so. It is information hiding, which helps a person break up the complexity of the method when reading it.

There are some questions about perf, but I think you've indicated that this function is only hit once on map load. Pls verify that assumption and I'll have no objection on those grounds. Second, timing the method should be very easy - we have a timer class that can be wired in without much effort, combined with a UI console that displays the output.

Last, this is core game engine code, I don't think the 'no-unit tests necessary' allowance applies here. The expectation is for new code to be built without adding significant technical debt. Missing regression tests is technical debt. This last point we probably should discuss in a different forum, and get the weigh in of everyone else on the topic.

DanVanAtta commented:
- This is a cool change
- Pls see last few paragraphs of last comment for next steps. Mainly just needs reviewer testing (we agreed that UI/functional changes would get functional testing by a reviewer).
- we need to decide on whether if tests are actually missing..

ron-murhammer commented:
@DanVanAtta - So I think the code is definitely more readable and maintainable now than it was when first submitted though I think we could have gotten to this point in a quicker/better fashion.

I went ahead and created a helper function for checking if the rectangle is contained in the territory and updated to base increments on width/height instead of font. I added in the timer but honestly I don't think we should leave it in as its too much output IMO. You could maybe move the timer up a few levels to capture the time it takes to draw whatever it selects on the screen instead of each territory but hard to even tell exactly the best place for that. It does show that essentially on load of WaW its takes under 500ms for the names to be placed (50% of the map or so) then as you scroll the rest are placed with no noticeable delay due to the name placement. Most method calls for territories are under 1ms with only occasion ones and very large ones greater than that but still <20ms. 

One interesting point is that it can be called multiple times in quick succession for the same territory though after it is drawn then its done. I think it has something to do with the territory being part of multiple tiles. But after you scroll around the entire map, it doesn't seem to get called again.

In regards to unit testing, we would have to make the method protected or public instead of private to test directly. This is also essentially an approximation algorithm so the only thing we can really test is a regression test where I run it to get the output then have the unit test verify what the code has just produced. Consider if we would decide to adjust the increments up or down, that would end up changing the output most likely just slightly but then the unit test fails. So then we would go and recopy the new results into the unit test. Since this is so low level and unlikely to really get changed much, a regression test doesn't have that much value. The only other thing I can think of is just test that the output is at least a rectangle contained in the territory (which primarily just tests the helper function portion which is actually the simple part).

A few closing thoughts:
- I have asked for unit test suggestions a few times but you have just continued to ask for unit tests which isn't really helpful. Providing the approach and input/output of what you recommend would be much more useful.
- I feel like the review has been sort of a couple of 'drive-by's' where you glance over things quickly and respond with a bunch of asks instead of taking the time to really read through the code, test the functionality, and pose well thought out questions. While I understand sometimes you have limited time, its probably better to focus on only glaring issues and posing some questions rather than demanding certain things be done some of which don't even make sense.

DanVanAtta commented on src/games/strategy/triplea/ui/screen/IDrawable.java:
```diff
@@ -231,6 +229,17 @@ private Rectangle getBestTerritoryNameRect(final MapData mapData, final Territor
 
+  private boolean isRectangleContainedInTerritory(final Rectangle rectangle, final Territory territory,
+      final MapData mapData) {
+    final List<Polygon> polygons = mapData.getPolygons(territory.getName());
```
Just going to point out, these two are equivalent:
```
    return polygons.stream().anyMatch(polygon -> polygon.contains(rectangle));
```

```
    for (final Polygon polygon : polygons) {
      if (polygon.contains(rectangle)) {
        return true;
      }
    }
    return false;
```

DanVanAtta commented on src/games/strategy/triplea/ui/screen/IDrawable.java:
```diff
@@ -187,41 +191,35 @@ public void draw(final Rectangle bounds, final GameData data, final Graphics2D g
+          final int score = verticalDistanceFromEdge * rectangle.width;
+          if (rectangle.width > nameWidth && score > maxScore) {
 
             // Check to make sure rectangle is contained in the territory
```
Comment can probably be removed now (but not a big deal)

DanVanAtta commented on src/games/strategy/triplea/ui/screen/IDrawable.java:
```diff
@@ -187,41 +191,35 @@ public void draw(final Rectangle bounds, final GameData data, final Graphics2D g
     final int centerY = territoryBounds.y + territoryBounds.height / 2;
-    final int minWidth = fontMetrics.stringWidth(territory.getName());
-    final int increment = fontMetrics.getAscent();
+    final int incrementX = (int) Math.ceil(territoryBounds.width / 20.0);
```
Good, you have the `Math.ceil` function :+1:

DanVanAtta commented on src/games/strategy/triplea/ui/screen/IDrawable.java:
```diff
@@ -187,41 +191,35 @@ public void draw(final Rectangle bounds, final GameData data, final Graphics2D g
-              }
-            }
-            if (isContained) {
+            if (isRectangleContainedInTerritory(rectangle, territory, mapData)) {
```
This helper may not have seemed so big, but I think it really helped. Helps keep for a separation of concerns :rabbit:

DanVanAtta commented:
### Some functional testing results
I checked whether you can do an early break from the outermost loop, here is a before:
![control_examples](https://cloud.githubusercontent.com/assets/12397753/14933995/9ee0f17e-0e56-11e6-8166-83756ab22917.png)

And after:
![experiment_examples](https://cloud.githubusercontent.com/assets/12397753/14933997/a320680a-0e56-11e6-8695-3c71d910bad5.png)

The results are different, so you are right @ron-murhammer about not being able to do an early break from the outer most loop. I thought I gave a pretty concise reasoning behind why I thought that was possible. Showing me the line of code that contradicted me would have been helpful. I'm still a bit curious where it is that I missed the dependency on the outermost loop variable.

### performance
Thank you for measuring it. I support not leaving the timer in. I took some measurements too and things were relatively fast. Something inconsistent caused territories to take 200ms to render. Could be a GC, a data structure resizing, any number of events.  There is a pretty well known upper bound to the number of loop iterations now as well, so I think we certainly have done our due diligence now.

### Remaining problems / missing logic

Text can overlap. It does seem like we will be able to control that more easily now thanks to this change. 

Here is an example of where the text still looks bad:
![mess_example](https://cloud.githubusercontent.com/assets/12397753/14934007/89071a9e-0e57-11e6-808b-21e4a1571927.png)

### unit testing

I'm sorry I wasn't more helpful in suggesting ways to unit test the functionality. I thought I had responded reasonably enough, but clearly it was not helpful. 

Thank you for the detailed response with regards to testing. You raised interesting points that I agree with. I agree a regression test would not be terribly helpful at this point.


### Retrospective on the Journey

- I also think things are in much better shape now than when this PR was initially opened. 
- I'm sorry the drive-by's caused so much friction. It is a bit of a double edged sword. On an initial pass I actively try to not understand the code too deeply. If something is complex enough it needs a deeper inspection, it's almost already worth calling out as it's already potentially too complex. If not then some extra guiding comments, careful design, and unit tests to lock down the complexity are expected. If I don't see that latter part, then it seems like a sign for trouble and hard to maintain code. The second edge to this sword is that I can do it too quickly, or while I'm a bit tired, and might be making a bad comment. So there is some improvement for me to make there, but hopefully the context helps. I can see now though that a 'drive-by' with a bajillion questions is overwhelming and can seem far too negative. It is something I can work on.
- For feedback responses, it seemed like sometimes it became a match of who was right, was getting personal. Clearly we were not helping each other, which is a shame since we really are on the same team (nobody else in the world is going to do this work for us). In cases where I misunderstood something, referencing line numbers or quoting sections of code could have been a succinct way to point out the misunderstandings. It was not helpful for me to simply be told that I did not understand.


So going forward:
- I can try to be more constructive with commentary, not pile-on quite so much.
- We should stop responding to comments we agree with, just let the comment go out of date and let the reviewer look at the commit diff.
- When getting feedback, our first reaction should be to update the code in whatever way possible to achieve some common ground, if there is any to be had. If there is none, it can be difficult to disagree cordially, but it really should simply lead to a discussion where get to learn, and/or think through a problem.

### Next steps for this PR
- We've agreed no unit test is needed
- It works
- The code by its nature is a bit complex, but the comments and refactoring updates have helped quite a bit. It's not readily clear what more could be done at this point.

No blockers on that list. Time to merge.

@ron-murhammer  thanks for hanging in there and the patience. I think I have felt very much the same you as you for some TripleA PRs that I had submitted. At the end though, I think we can say that this clearly seems ready now, and it's been improved quite a bit.
